### DISCUSSION TITLE
Why is Scheduler#scheduler not interruptible?

### DISCUSSION THREAD

**Discussion by alexandru** [lampepfl/gears#65] [General]
Looking at the behaviour exemplified by this test:

```scala
  test("schedule cancellation doesn't abort inner code") {
    Async.blocking:
      var bodyRan = false
      val fut = Promise[Unit]()
      val cancellable = Async.current.scheduler.schedule(
        50.milliseconds,
        () =>
          fut.complete(Success(()))
          Async.blocking:
            sleep(500)
            bodyRan = true
      )

      // cancel after body started running
      fut.await
      cancellable.cancel()

      sleep(1000)

      assert(bodyRan)
  }
```

What's the reason for this contract? Cancelling a scheduled task, while running, is a fair use-case. Java exposes a [mayInterruptIfRunning](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#cancel-boolean-) boolean flag, for such cases, however, in its absence, the better default would be to interrupt if running. Because you can make any scheduled code to be non-interruptible.

**Comment by natsukagami** [comment]
In gears we are currently settling with *not* requiring Scheduler tasks to be fully interruptible. This cannot be done e.g. in Scala Native without tracking the underlying running thread and try to send an interrupt signal to it.

I think for now, while we're looking at preemption patterns, that this is not something I'd be eager to expose.
